gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -132,29 +149,113 @@ public static void renamePath(FileSystem fs, Path oldName, Path newName) throws
   }
 
   /**
-   * A wrapper around {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)}
-   * which throws {@link IOException}
-   * if {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} returns false.
+   * Moves a src {@link Path} from a srcFs {@link FileSystem} to a dst {@link Path} on a dstFs {@link FileSystem}. If
+   * the srcFs and the dstFs have the same scheme, and neither of them or S3 schemes, then the {@link Path} is simply
+   * renamed. Otherwise, the data is from the src {@link Path} to the dst {@link Path}. So this method can handle copying
+   * data between different {@link FileSystem} implementations.
+   *
+   * @param srcFs the source {@link FileSystem} where the src {@link Path} exists
+   * @param src the source {@link Path} which will me moved
+   * @param dstFs the destination {@link FileSystem} where the dst {@link Path} should be created
+   * @param dst the {@link Path} to move data to
    */
-  public static void copyPath(FileSystem fs, Path src, Path dst) throws IOException {
-    if (!FileUtil.copy(fs, src, fs, dst, false, fs.getConf())) {
-      throw new IOException(String.format("Failed to copy %s to %s", src, dst));
+  public static void movePath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, Configuration conf)
+      throws IOException {
+
+    if (srcFs.getUri().getScheme().equalsIgnoreCase(dstFs.getUri().getScheme()) && !S3_SCHEMES
+        .contains(srcFs.getUri().getScheme()) && !S3_SCHEMES.contains(dstFs.getUri().getScheme())) {
+      renamePath(srcFs, src, dst);
+    } else {
+      copyPath(srcFs, src, dstFs, dst, true, false, conf);
     }
   }
 
   /**
-   * A wrapper around {@link HadoopUtils#renamePath(FileSystem, Path, Path)} and
-   * {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} which will rename the path
-   * if the src and dst filesystems are the same; otherwise, the src file will be moved the the dst filesystem.
-   * An {@link IOException} if {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} returns false.
+   * Copies a src {@link Path} from a srcFs {@link FileSystem} to a dst {@link Path} on a dstFs {@link FileSystem}. If
+   * either the srcFs or dstFs are S3 {@link FileSystem}s (as dictated by {@link #S3_SCHEMES}) then data is directly
+   * copied from the src to the dst. Otherwise data is first copied to a tmp {@link Path}, which is then renamed to the
+   * dst.
+   *
+   * @param srcFs the source {@link FileSystem} where the src {@link Path} exists
+   * @param src the {@link Path} to copy from the source {@link FileSystem}
+   * @param dstFs the destination {@link FileSystem} where the dst {@link Path} should be created
+   * @param dst the {@link Path} to copy data to
+   * @param tmp the temporary {@link Path} to use when copying data
```

Comment:
Yes, it seems that the HDFS FileSystem originally tried to do something similar and had to make it configurable: https://issues.apache.org/jira/browse/HADOOP-11281

A `hdfs dfs -cp [src] [dst]` also uses a tmp file. It first copies data to `[dst]._COPYING_` and then does a rename. Seems there is also an option to disable this so that data is directly copied from the `src` to the `dst` (useful for S3). Unfortunately, the code doesn't seem to be publicly available.